source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  skip_compression_ = !isAcceptEncodingAllowed(headers);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && !hasCacheControlNoTransform(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding() && isEtagAllowed(headers)) {
+
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data = data.length();
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::hasCacheControlNoTransform(HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return StringUtil::findToken(cache_control->value().c_str(), ",",
+                                 Http::Headers::get().CacheControlValues.NoTransform.c_str());
+  }
+  return false;
+}
+
+// https://tools.ietf.org/html/rfc7231#page-41
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    for (const auto token :
+         StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", false)) {
+      const auto value = StringUtil::trim(StringUtil::cropRight(token, ";", false));
```

Comment:
the false args here are all for trim_whitespace right?  Can you comment that, e.g.

cropRight(token, ";", false /* trim_whitespace */);

And also...are you sure you don't want to trim whitespace in any of the cases?

Reply:
We need to trim both sides however, the trim_whitespace argument in crop functions only allows doing `ltrim` or `rtrim` as suggested in previous code review for the utility library. I'd rather do one full trim after the token has been split and cropped.

Reply:
My bad...missed the fact that you were trimming it.  Looks great.